src/app_modes/viewport.rs

```diff
@@ -0,0 +1,235 @@
+use crate::app_modes::{input, AppMode, Drawable};
+use crate::footprint::get_current_footprint;
+use crate::listeners::Listeners;
+use crate::transformation;
+use std::sync::Arc;
+use tui::backend::Backend;
+use tui::layout::{Constraint, Layout};
+use tui::style::Color;
+use tui::widgets::canvas::{Canvas, Context, Line, Points};
+use tui::widgets::{Block, Borders};
+use tui::Frame;
+
+pub trait UseViewport: AppMode {
+    fn draw_in_viewport(&self, ctx: &mut Context);
+    fn x_bounds(&self) -> [f64; 2];
+    fn y_bounds(&self) -> [f64; 2];
+    fn info(&self) -> String;
+}
+
+impl<B: Backend, T: UseViewport> Drawable<B> for T {
+    fn draw(&self, f: &mut Frame<B>) {
+        let chunks = Layout::default()
+            .constraints([Constraint::Percentage(100)].as_ref())
+            .split(f.size());
+
+        let canvas = Canvas::default()
+            .block(
+                Block::default()
+                    .title(format!("{} - {}", self.get_name(), self.info()))
+                    .borders(Borders::NONE),
+            )
+            .x_bounds(self.x_bounds())
+            .y_bounds(self.y_bounds())
+            .paint(|ctx| {
+                self.draw_in_viewport(ctx);
+            });
+        f.render_widget(canvas, chunks[0]);
+    }
+}
+
+pub struct Viewport {
+    pub static_frame: String,
+    pub robot_frame: String,
+    pub tf_listener: Arc<rustros_tf::TfListener>,
+    pub initial_bounds: Vec<f64>,
+    pub footprint: Vec<(f64, f64)>,
+    pub axis_length: f64,
+    pub zoom: f64,
+    pub zoom_factor: f64,
+    pub terminal_size: (u16, u16),
+    pub listeners: Listeners, // TODO split properly config and listeners
+}
+
+impl Viewport {
```

Comment:
phrasing it differently, viewing only  should be a mode of its own and be default ;)

Reply:
I wanted the viewport to be more abstract to have the other modes cleaner. A view mode was my first thought but it's pretty redundant with the send pose since now the camera tracks the ghost rather than the robot.

If you think it makes sense I'll make a view mode by wrapping the viewport, but I would not expose the viewport itself.

Reply:
I'm also pro extra mode for just viewing which is the default. it prevents users from doing stupid stuff when they drop in